### ISSUE TITLE
Display file information in treemap

### ISSUE THREAD

**Issue by alanbernstein** [shundhammer/qdirstat#73]
I am one of those occasional treemap users that prefer to see path information in the treemap display, and I wonder if this is feasible with qdirstat.

I can see you have no interest in supporting such options (https://github.com/shundhammer/qdirstat/issues/57, https://github.com/shundhammer/qdirstat/issues/36). Instead of asking you to do so, I instead ask for any advice you might have for someone trying to implement this in (a fork of) qdirstat.

**Comment by shundhammer** [comment]
So, you want advice how to do something that I already explained cannot be done properly?

Just look at some typical treemaps:

![work](https://user-images.githubusercontent.com/11538225/35241977-91ecf4cc-ffb8-11e7-96b8-6952ae9285c6.png)

![usr](https://user-images.githubusercontent.com/11538225/35241989-99813914-ffb8-11e7-80b5-407a80fb3655.png)

![src](https://user-images.githubusercontent.com/11538225/35241995-9ec946aa-ffb8-11e7-9fcf-7e6bb4592cb7.png)

How many treemap tiles do you see that you could fit any path information into, even if you used a really tiny font?

Look at an original _Space Sniffer_ screenshot:

![spacesniffer](https://user-images.githubusercontent.com/11538225/35242160-129415ce-ffb9-11e7-8d3a-bfe927319894.png)

Even with that very small number of files you can only read the filename (without any path information!) and the file size in very few cases.

**Comment by shundhammer** [comment]
So, even with a really tiny font size, it just won't work, much less for displaying a full-sized Linux path.

Sure, you could change the treemap tiles' rendering - either render a string as an overlay, preferably with a rectangle with a slightly darker or lighter background and some (60%?) transparency in the alpha channel. You could give each treemap tile another QGraphicsItem child where you do this. You'd still have to clip the string in almost all cases, though.

Not only would it get you a very ugly visual appearance, it would also be in vain in almost all cases since you _just can't read the stuff_ because it's cut off, very often even in both dimensions.

So, what's the point? Adding stuff for the sake of adding stuff? Drowning the user in information which is just illegible garbage in almost all cases? Where is benefit for the user?

**Comment by alanbernstein** [comment]
Simply put, the Space Sniffer approach is useful to me, and the "cushioned" display with no text is not useful to me. I can elaborate on my specific use case and workflow if you like. 

I have written my own (very rough) space visualization tool, displaying information exactly as I want it. Here is a more representative screenshot of this type of display:

![screen shot 2018-01-22 at 8 27 23 pm](https://user-images.githubusercontent.com/16123214/35255343-3768c2f0-ffb4-11e7-9c78-32f27898d4e5.png)

One notable difference is that color indicates depth (very much like the old Spacemonger approach). I also use a monitor with much higher resolution than the screenshots you shared (yes, I can and do read the text at that size). Ideally, I would prefer to leverage a more mature project like yours, rather than continue to hack features onto my application.

Long ago, I read [this paper](https://www.win.tue.nl/~vanwijk/ctm.pdf) presenting the cushion approach. In the section motivating the work, several "drawbacks" of other styles are presented. At least one of these drawbacks is outdated, and none of them appear to be based on any objective science. Personally, I disagree with all of them, and I am also unable to extract any meaningful information from the cushioned display. You, like the authors of that paper, present these drawbacks as objective truths, without much to support such a claim.

I understand you are not interested in this, nor do you approve of it. I am only asking for a few pointers for getting started working with your codebase. `QGraphicsItem` is a good keyword to know, but it's not quite enough. If that is too much to ask, so be it. Thanks for your time.

**Comment by shundhammer** [comment]
Well, did you read the existing documentation in the two classes with "treemap" in the name?

You will be hard-pressed to find any piece of software (let alone Open Source software) with more exhaustive documentation.

Still, that part of QDirStat is (as one might expect) among the most complex ones (the other one being the Qt data model for the tree), so you will have to dig deep into it to understand what's going on. And you will need to read the documentation for the underlying Qt classes, of course (QGraphicsView, QGraphicsScene, QGraphicsItem) which all by itself is not trivial.

**Comment by shundhammer** [comment]
As for "objective science" that you mentioned - isn't it obvious to you that not only is your screenshot *huge* (4064 x 2570 pixels), your treemap also violates a basic principle of treemaps: That the area of each individual tile corresponds to the associated value. Since you need some pixels for all those borders and margins around each subdivision, the deeper you get into the tree, the more they accumulate, which of course leaves considerably less space for the leaf nodes (the files) which actually have that value (the file size), so a file of, say, 4 MB two directory levels down gets more space (has a larger area) than a file of the same size 8 or more directory levels down? You cannot compare those treemap tiles anymore because of this, so the visualization is misleading.

Also, that huge treemap in your example does not leave any screen space for a connected view such as the tree, not to mention things like a menu bar or a tool bar.

Still, the labels that are not cut off and thus are readable are very few - just a handful of the larger ones.

Since you picked a source directory those all happen to be git "pack" files - you don't see the full path (which I thought was the whole point of this exercise), only a meaningless name "pack-0a992dc...", i.e. some meaningless hex number.

So, a user can recognize that it's the git pack files that consume most space in this example. Fine; this could have been achieved much easier by providing a separate MIME type with a special color for those files - which is something you can already easily do in the existing QDirStat.

As predicted, even with this _huge_ treemap that few (if any) users will ever see because most don't have a 4k display, 95% or more of all labels are cut off to the point of being worthless to the user.

But any MIME type information gets lost on the way; there is no way to see if it's your photo collection, your videos, your MP3s or any other file type consuming so much space.

So you trade information that does have a meaning for the user against labels that are illegible in almost all cases (thus are useless to the user).

What again was the point of all this?

So, things that are so obvious are not sufficiently "objective science" enough?

**Comment by shundhammer** [comment]
Having said all this, I might still be inclined to agree to have such a kind of view as an alternative the user can switch to in QDirStat.

It would complicate the part of the code that is already the most complicated even more, yes, but I'd be willing to bite that bullet if there was any worthwhile return value for the user.

But unfortunately, it's not just the treemap rendering; this can already be changed to flat colors with existing configuration options (https://github.com/shundhammer/qdirstat/issues/36):

![flat-treemap](https://user-images.githubusercontent.com/11538225/35269560-a724e7ea-002c-11e8-8afe-569679ec7d29.png)

But having all those frames and labels for each directory would break the layout algorithm, so this wouldn't work that easily.

At the very minimum, this would mean subclassing the TreemapView and the TreemapTile classes from a common base class for each so at least all the user interaction (clicking, resizing, context menu with cleanup actions) might be shared between the existing approach and that new one, and of course the communication with the attached tree view.

This all sounds like a lot of work. There would have to be some real value for the end user behind it to justify that, and I just don't see that.

**Comment by alanbernstein** [comment]
I have not yet looked at the code at all; asking a more knowledgeable person for a small bit of guidance is often the best way to get started. I don't doubt that it will be complicated and messy, and I may not make much progress on it. Without those tips from you, I probably would never try.

I'll reiterate that I have no expectation of you working on this, or even accepting any work that I might do. Nor do I have any desire to change your mind, but you've asked so I'll reply:

Yes, the screenshot is huge, I'm not sure why you point to that as a problem. I have been waiting for truly high-resolution monitors since I started working with computers, and of course I'm taking advantage of them now that they exist. I find full-screen treemaps to be one of the best ways to do that. I am not going to speculate on the proportion of users that do this, but 4K displays are not in themselves extraordinary. On the contrary, I find the small size of your screenshots surprising. Do you really use QDirStat in a window that is smaller than 1000x800 pixels? Why? Do you think that is _more_ representative than the screenshot I shared?

I have had little need for a menu bar or tree display, as almost all of the information I care about is present in the treemap. Of course those widgets would be a nice addition, but they would take a small fraction of the available space. I tend to use the treemap in conjunction with a terminal where, for example, tab-completion makes it easy to ignore details like the full 40-character git pack filename. This is part of the reason I generally don't need a full path, but rather just enough to identify a file.

When I use one of these tools, it is because I am asking the question "what is taking up space on this drive?" I'm sure this goal is shared between us, but from there on it seems that we diverge - it's not clear to me what you yourself might do with that knowledge. Once I have that, my next step is usually to reclaim space on the drive, by finding objects that can be deleted immediately. To that end, I look for the largest objects that can be deleted all at once. These might be project directories, media files, data files, obsolete software - anything really - but, again, the goal is to *find the largest single objects that can be deleted immediately*.

This has a few implications:

First, I overwhelmingly focus on the largest, self-contained objects in the display. With the ability to zoom the display arbitrarily, it is always possible to re-render the treemap such that the objects that I care about are the largest objects.

Second, seeing space use by (for example) MIME type, scattered across the whole treemap, is generally not helpful to me. I can distinguish photos/videos/mp3s by reading the top-level labels, because my media files are organized as such, but of course that is not always the case. Say I could learn that all git pack files use 2GB out of 20GB of source code. What would I do with that knowledge? I'm not going to delete all of them at once. I'm not going to reduce or otherwise modify any of them on their own. The only meaningful action I can take with respect to these files is to delete individual repository directories in full. The focus is on *repos*, not on *file types*. 

With these points in mind:

The treemap is useful as an approximate visualization where order of magnitude and relative size are most important, while precise size ratios are less so. Anyway, the error introduced by the margins is small for the larger objects, especially when the canvas is large. This is why the labels are not a problem. You mention that 95% of the labels are cut off, and this is true, by count. But 95% of the _space_ is captured by fully visible labels. The ones that are cut off - in the smallest, deepest nodes - don't matter to me. I trade exactness for practicality. Why even display the little stuff then? It's just an arbitrary threshold, it doesn't impact me, so I don't bother to tinker with it.

"It's obvious" is not objective science. It's equally obvious, to me, that my preferences are better for my use case. It's the best solution I've found to the problem I have, but I don't think it's perfect, and I'd be happy to improve it. Maybe you'll point out where I'm wrong, and I'll learn how to use QDirStat productively in its current form. Or maybe I simply have a different use case than you and 99% of your users, and we can accept that and live with it.

**Comment by shundhammer** [comment]
So you don't want anything of what makes QDirStat what it is: Not the tree, not the menus and buttons, thus also not the cleanup actions. And you also want to change everything about the treemap.

So, why bother with QDirStat at all? What's left for you?

**Comment by shundhammer** [comment]
As for how people use the treemap and everything else: The treemap as QDirStat displays it is great at two things:

- Pointing out files that are disproportionally large.
- Giving a much simplified impression about certain MIME types that consume a lot of disk space and possibly also how they are grouped.

In some cases, this means that users spot that huge ISO image that they downloaded a while ago and forgot about it, but since it stands out like a sore thumb in the treemap they can decide if they want to get rid of it (and do that from _within_ QDirStat with a few mouse clicks).

In other cases, they might realize that certain shared libs are really large, but that doesn't mean that they can or want to get rid of them; they will probably need them to keep some of their software working. So they get the information, but they cannot act upon it.

But that's roughly the limits of the treemap. It is poor at showing accumulated sums of directories; this is what the tree view does a lot better.

In your example, the information that it's really some of those git pack files that are so large is just as useless as in the above example the information that some shared libs are so large: You can't just get rid of them.

What you can do, however, is to decide if you really still need that entire git checkout directory with all the sources; you might have checked it out some time ago to see how some software does things, and you might not need or want that entire directory anymore.

Again, it is the tree view that gives you that information, and it is the tree view from where you can easily delete that directory. Or you might decide to at least execute `make clean` or `git clean -dfx` there so you can still keep the sources, but not all the built object files and binaries.

This is what QDirStat is all about. And to facilitate that, it has a "selection model" that exchanges information between the attached view widgets (the tree view and the treemap view) which items are selected and which one has the keyboard focus etc.; this is where a lot of application logic and know-how sits, and accordingly, this part is also some of the more complex code. Parts of that code are also in the two view classes where the signals from the selection model are received and acted upon (highlight selected items, enable or disable cleanup actions based on what type of object is selected and how many of them).

So, if you don't want the tree view, there is no point in keeping all that around; it would only be a lot of code (and pretty complex code, too) to be maintained for no good reason.

And if you'd want to change all the rendering and all the layout logic of the treemap, what's left of QDirStat's treemap classes that would be useful for you? It would just be in the way.

So, the only thing left over would be the directory tree classes with its corresponding item classes and the directory reading logic with its job queue. That's not really much (and it's not exactly rocket science either).

Basically, what you are up to is to recreate the old _SequoiaView_ from the TU Eindhoven people who wrote those papers about shaded treemaps, but without the shading: An app to display a treemap (but with different rendering) with maybe the ability to delete individual files, but little more.

So, again, why bother starting this based on QDirStat?

**Comment by shundhammer** [comment]
> "It's obvious" is not objective science. It's equally obvious, to me, that my preferences are better for my use case. 

This is bullshit argumentation. Sorry. I pointed out _why_ and _what_ while you insist on just personal impressions.

**Comment by alanbernstein** [comment]
I'm just looking to improve my own workflow. The display format is the most important aspect to me, which is why I took on that task first. The tool that I created to do that is not a proper GUI application, it creates more-or-less static images that are not easy to navigate. I am not experienced with GUI development, so if I want to improve things for myself, it will be a large investment whether I add features to my own tool, or figure out how to adapt another one. Yours seems like a decent choice for that.
